[develop] Add verification of snowfall accumulation

willmayfield opened the pull request:
## DESCRIPTION OF CHANGES: 
This PR completes the final deliverable of the DTC UFS SRW project to include verification support of new variables in the workflow via MET. This PR adds relevant METplus configuration files and modifies various scripts to allow for verification via the NOHRSC gridded accumulated snowfall product. The forecast variable created by post is 'TSNOWP' for 6-and 24-hour snowfall accumulation with a fixed density, which will be verified on a 6- and 12-hour cadence, respectively. When the GSL variable-density snowfall product is included in post in the future, the variable in the METplus configuration files may be changed to 'ASNOW'. 

Currently this addition does not turn on accumulated snowfall verification by default. This is because NOHRSC observations were not located on NOAA HPSS previous to March 2020, as well as the idea that many non-winter cases likely will not desire to run the additional verification tasks. This PR achieves removal of snowfall verification tasks by omitting it by default from the VX_FIELDS variable. Therefore, to turn on snowfall verification, users may simply include VX_FIELDS in their configuration with "ASNOW" added.

This PR also turns off all “ncpairs” file creation in METplus configuration files. These files represent the majority of the filespace footprint for METplus outputs, and are not necessary to produce verification statistical output.

### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## TESTS CONDUCTED: 
MET_verification and MET_ensemble_verification tests were run without error. Note that features were not tested as a part of these because the new "ASNOW" tasks are not run by default, as noted above. A winter case (20220202) was run for a 36 hour ensemble forecast that included "ASNOW" in VX_FIELDS. The experiment directory on hera is in /scratch2/BMC/fv3lam/mayfield/srw_vx/snowfall_vx/expt_dirs/test_community_nohrsc. METplus stat files were reviewed for meaningful statistics, and “ncpairs” files were turned on and reviewed to show that both observations and forecasts as processed by METplus are verifiable.

- [X] hera.intel
- [ ] orion.intel
- [ ] cheyenne.intel
- [ ] cheyenne.gnu
- [ ] gaea.intel
- [ ] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [ ] fundamental test suite
- [ ] comprehensive tests (specify *which* if a subset was used)

## DEPENDENCIES:
None

## DOCUMENTATION:
Documentation was updated to include additional information about NOHRSC observation data in verification sections.

## ISSUE: 
None

## CHECKLIST
- [X] My code follows the style guidelines in the Contributor's Guide
- [X] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] New and existing tests pass with my changes
- 
## REQUEST SPECIFIC REVIEWERS:
@gsketefian @mkavulich

willmayfield committed:
Initial work to add NOHRSC snowfall accumulation obs pulling and pcpcombine

willmayfield committed:
new files for snowfall accumulation obs pulling and pcpcombine

willmayfield committed:
new files for snowfall accumulation obs pulling and pcpcombine

willmayfield committed:
metplus conf files for pcp_combine and gridstat related to accumulated snowfall

willmayfield committed:
metplus conf files for pcp_combine and gridstat related to accumulated snowfall

willmayfield committed:
GridStat vx for snowfall accumulation

willmayfield committed:
GridStat vx for snowfall accumulation

willmayfield committed:
Changes for ensemble verification of snowfall accumulation

willmayfield committed:
Changes for ensemble verification of snowfall accumulation

willmayfield committed:
documentation and cleanup for snowfall vx

willmayfield committed:
documentation and cleanup for snowfall vx

willmayfield committed:
documentation formatting

willmayfield committed:
documentation formatting

willmayfield committed:
clean whitespace

willmayfield committed:
clean whitespace

willmayfield committed:
documentation updates and code cleanup

willmayfield committed:
documentation updates and code cleanup

willmayfield committed:
Resolve conflicts

willmayfield committed:
Merge branch 'feature/snowfall_vx' of https://github.com/willmayfield/ufs-srweather-app into feature/snowfall_vx

willmayfield committed:
Remove ncpairs output turned on by default

MichaelLueken commented on docs/UsersGuide/source/RunSRW.rst:
```diff
@@ -722,12 +724,14 @@ Users who have already staged the observation data needed for METplus (i.e., the
 
    platform:
       CCPA_OBS_DIR: /path/to/UFS_SRW_App/develop/obs_data/ccpa/proc
+      NOHRSC_OBS_DIR: /path/to/UFS_SRW_App/develop/obs_data/ccpa/proc
```
Should the directory path be:

`/path/to/UFS_SRW_App/develop/obs_data/nohrsc/proc`, or is the current addition correct?

MichaelLueken commented on scripts/exregional_get_obs_nohrsc.sh:
```diff
@@ -0,0 +1,135 @@
+#
+# This script retrieves and organizes the NOHRSC data into a more intuitive structure:
+# A valid YYYYMMDD directory is created, and all files for the valid day are placed within the directory.
+# NOTE: Accumulation is currently hardcoded to 06h and 24h which are aavailable every 6 and 12 hour, respectively..
```
`available` rather than `aavailable` and please remove second period at the end of the sentence.

```suggestion
# NOTE: Accumulation is currently hardcoded to 06h and 24h which are available every 6 and 12 hour, respectively.
```

MichaelLueken reviewed (commented):
@willmayfield I have gone over your changes in this PR and overall everything looks pretty good.  I have noted a few spelling/grammatical issues and had a question regarding the `NOHRSC_OBS_DIR` path in `RunSRW.rst`, but otherwise, everything else looks okay.

I was able to successfully run all of the verification WE2E tests on Hera, as well as the fundamental WE2E tests.  I was even able to test the newly added ASNOW option.  I do have a question regarding this.

The `metatask_PcpCombine_fcst_APCP_all_accums_all_mems` entry correctly sets `ACCUM_HH` to `01`, `03`, and `06` for `FCST_LEN_HRS` = 6.  I noted that the only way that I could get `metatask_PcpCombine_fcst_ASNOW_all_accums_all_mems` to set `ACCUM_HH` to `06` was by setting `FCST_LEN_HRS` to a value of 24 or greater.  Is this correct, or should I be able to set `FCST_LEN_HRS` = 6 and still have `ACCUM_HH` set to `06`?  If this is what is supposed to happen, would you mind running a test with `FCST_LEN_HRS` = 6 and seeing whether `ACCUM_HH` is properly set?

Once the above question has been answered, I think I can move forward with approving these changes.

willmayfield committed:
fix typos and duplicated line

willmayfield commented:
@MichaelLueken Thank you for your review. I fixed the typos and another duplicate line I noticed in default_workflow.yaml. 

Good catch for the issue when forecast length is less than 24h. It seems when only one element of the verification.VX_ASNOW_ACCUMS_HRS list is used, it propagates ACCUM_HH without the 2-digit formatting, but I have no idea why, as it should happen in the loop in verify_pre.yaml:
`ACCUM_HH: '{% for ah in verification.VX_ASNOW_ACCUMS_HRS %}{% if workflow.FCST_LEN_HRS >= ah %}{{ "%02d " % ah }}{% endif %}{% endfor %}'`
 I also checked that for a 2 hour forecast length, the APCP tasks also fail for the same reason. Checking FV3LAM_wflow.xml, I see the following:
For a 5 hour forecast length:
```
<metatask name="PcpCombine_fcst_APCP_all_accums_all_mems" >
  <var name="ACCUM_HH">01 03 </var>
```

For a 2 hour forecast length:
```
<metatask name="PcpCombine_fcst_APCP_all_accums_all_mems" >
  <var name="ACCUM_HH">1</var> 
```
(the workflow also fails because the ACCUM_HH var is empty for PcpCombine_fcst_ASNOW_all_accums_all_mems, so I will try to find a way to handle that as well, probably by removing those tasks with some logic in setup.py)

Maybe @gsketefian or @mkavulich can offer any ideas of why it isn't doing the 2-digit formatting?

gsketefian commented on docs/UsersGuide/source/ConfigWorkflow.rst:
```diff
@@ -151,23 +151,31 @@ METplus Parameters
 
 ``CCPA_OBS_DIR``: (Default: "")
-   User-specified location of top-level directory where CCPA hourly precipitation files used by METplus are located. This parameter needs to be set for both user-provided observations and for observations that are retrieved from the NOAA :term:`HPSS` (if the user has access) via the ``GET_OBS_CCPA`` task. (This task is activated in the workflow by using the taskgroup file ``parm/wflow/verify.yaml ...
+   User-specified location of top-level directory where CCPA hourly precipitation files used by METplus are located. This parameter needs to be set for both user-provided observations and for observations that are retrieved from the NOAA :term:`HPSS` (if the user has access) via the ``GET_OBS_CCPA`` task. (This task is activated in the workflow by using the taskgroup file ``parm/wflow/verify_pre. ...
```
@willmayfield Suggestion:  To be clearer, instead of saying "This task is activated...", say something like "The workflow task that pulls CCPA observations is activated..." since it's not clear what "This task" refers to.

gsketefian replied:
Change the task name to lowercase, i.e. `GET_OBS_CCPA` --> `get_obs_ccpa`.

gsketefian commented on docs/UsersGuide/source/ConfigWorkflow.rst:
```diff
@@ -151,23 +151,31 @@ METplus Parameters
+
 ``MRMS_OBS_DIR``: (Default: "")
-   User-specified location of top-level directory where MRMS composite reflectivity files used by METplus are located. This parameter needs to be set for both user-provided observations and for observations that are retrieved from the NOAA :term:`HPSS` (if the user has access) via the ``GET_OBS_MRMS`` task (activated in the workflow automatically when using the taskgroup file ``parm/wflow/verify. ...
+   User-specified location of top-level directory where MRMS composite reflectivity files used by METplus are located. This parameter needs to be set for both user-provided observations and for observations that are retrieved from the NOAA :term:`HPSS` (if the user has access) via the ``GET_OBS_MRMS`` task (activated in the workflow automatically when using the taskgroup file ``parm/wflow/verify_ ...
```
Change the task name to lowercase, i.e. `GET_OBS_MRMS` --> `get_obs_mrms`.

gsketefian commented on docs/UsersGuide/source/ConfigWorkflow.rst:
```diff
@@ -151,23 +151,31 @@ METplus Parameters
 
 ``NDAS_OBS_DIR``: (Default: "")
-   User-specified location of the top-level directory where NDAS prepbufr files used by METplus are located. This parameter needs to be set for both user-provided observations and for observations that are retrieved from the NOAA :term:`HPSS` (if the user has access) via the ``GET_OBS_NDAS`` task (activated in the workflow automatically when using the taskgroup file ``parm/wflow/verify.yaml``). W ...
+   User-specified location of the top-level directory where NDAS prepbufr files used by METplus are located. This parameter needs to be set for both user-provided observations and for observations that are retrieved from the NOAA :term:`HPSS` (if the user has access) via the ``GET_OBS_NDAS`` task (activated in the workflow automatically when using the taskgroup file ``parm/wflow/verify_pre.yaml`` ...
```
Change the task name to lowercase, i.e. `GET_OBS_NDAS` --> `get_obs_ndas`.

gsketefian commented on docs/UsersGuide/source/ConfigWorkflow.rst:
```diff
@@ -151,23 +151,31 @@ METplus Parameters
       There is a problem with the valid time in the metadata for files valid from 19 - 00 UTC (i.e., files under the "00" directory). The script to pull the CCPA data from the NOAA HPSS (``scripts/exregional_get_obs_ccpa.sh``) has an example of how to account for this and organize the data into a more intuitive format. When a fix is provided, it will be accounted for in the ``exregional_get_obs_c ...
 
+``NOHRSC_OBS_DIR``: (Default: "")
+   User-specified location of top-level directory where NOHRSC 06- and 24-hour snowfall accumulation files (available every 6 and 12 hours respectively) used by METplus are located. This parameter needs to be set for both user-provided observations and for observations that are retrieved from the NOAA :term:`HPSS` (if the user has access) via the ``GET_OBS_NOHRSC`` task. (This task is activated i ...
```
Change the task name to lowercase, i.e. `GET_OBS_NOHRSC` --> `get_obs_nohrsc`.

gsketefian commented on docs/UsersGuide/source/RunSRW.rst:
```diff
@@ -827,18 +831,24 @@ In addition to the baseline tasks described in :numref:`Table %s <WorkflowTasksT
    | **Workflow Task**     | **Task Description**                                       |
    +=======================+============================================================+
    | GET_OBS_CCPA          | Retrieves and organizes hourly :term:`CCPA` data from NOAA |
```
@willmayfield @gspetro This is not directly related to this PR, but we will have to update the names of the vx and related tasks in this table, e.g. `GET_OBS_CCPA` becomes `get_obs_ccpa`, etc.  I think I'll do that as part of Issue #630.

gspetro-NOAA replied:
@gsketefian I'm actually working on updates now! The code is in my text/ug-updates branch.
RTD version here: https://srw-ug.readthedocs.io/en/text-ug-updates/BuildingRunningTesting/RunSRW.html#vxworkflowtaskstable 
I'm generally not touching much of the VX stuff, but since I'm updating the run chapter, I did get update some of the info there.

gspetro-NOAA replied:
One thing I was trying to decide was whether to name the (meta)tasks based on what is in the YAML config file or based on what the Rocoto/log output is. I went with what is in the YAML because that is the name users may actually add/change/remove in their config file.

gsketefian commented on parm/metplus/GridStat_APCPgt01h.conf:
```diff
@@ -257,13 +257,13 @@ GRID_STAT_OUTPUT_FLAG_NBRCNT = STAT
 #GRID_STAT_NC_PAIRS_VAR_NAME =
-GRID_STAT_NC_PAIRS_FLAG_LATLON = TRUE
-GRID_STAT_NC_PAIRS_FLAG_RAW = TRUE
+GRID_STAT_NC_PAIRS_FLAG_LATLON = FALSE
```
Just curious why we're turning off `GRID_STAT_NC_PAIRS_FLAG_LATLON`, `GRID_STAT_NC_PAIRS_FLAG_RAW`, and `GRID_STAT_NC_PAIRS_FLAG_NBRHD` (see [here](https://met.readthedocs.io/en/latest/Users_Guide/grid-stat.html?highlight=nc_pairs_flag%20latlon)).

gsketefian commented on parm/metplus/GridStat_APCP01h.conf:
```diff
@@ -246,13 +246,13 @@ GRID_STAT_OUTPUT_FLAG_NBRCNT = STAT
 #GRID_STAT_NC_PAIRS_VAR_NAME =
-GRID_STAT_NC_PAIRS_FLAG_LATLON = TRUE
-GRID_STAT_NC_PAIRS_FLAG_RAW = TRUE
+GRID_STAT_NC_PAIRS_FLAG_LATLON = FALSE
```
Just curious why we're turning off `GRID_STAT_NC_PAIRS_FLAG_LATLON`, `GRID_STAT_NC_PAIRS_FLAG_RAW`, and `GRID_STAT_NC_PAIRS_FLAG_NBRHD`.  See [here](https://met.readthedocs.io/en/latest/Users_Guide/grid-stat.html?highlight=nc_pairs_flag%20latlon) for descriptions of these options (or their conterparts in MET).

willmayfield replied:
@gsketefian I discussed this with @michelleharrold and we decided it would be best to have these turned off by default. The ncpairs files are useful for checking that verification fields are created correctly, but the files are not needed in order to create the verification statistics. They can be rather large files as well, so it will help with reducing the footprint of the verification WE2E tests.

gsketefian commented on scripts/exregional_get_obs_nohrsc.sh:
```diff
@@ -0,0 +1,135 @@
+#-----------------------------------------------------------------------
+#
+
+set -x
```
Probably should remove this `set -x` since it adds a lot of clutter to the log file.  User can add it when debugging (or you can put this in an if-statement on the `DEBUG` or similar flag; or maybe that already exists in `$USHdir/preamble.sh`).

gsketefian commented on scripts/exregional_run_met_genensprod_or_ensemblestat.sh:
```diff
@@ -180,16 +183,22 @@ if [ "${grid_or_point}" = "grid" ]; then
+        FCST_INPUT_DIR="${vx_fcst_input_basedir}"
+        ;;
+      "ASNOW")
+        OBS_INPUT_FN_TEMPLATE="${OBS_NOHRSC_ASNOW_FN_TEMPLATE}"
```
@willmayfield I will try to run this PR myself to understand better what's going on, but can you say why `FCST_INPUT_DIR` for `ASNOW` is set to `vx_output_basedir`?  Is there a vx pre-processing task (like `PCP_COMBINE` for `APCPgt01h`) that first needs to run and that places output in `vx_output_basedir`?

willmayfield replied:
@gsketefian Yes, pcpcombine_fcst runs for ASNOW prior to this task.

gsketefian replied:
Oh, just saw that the answer to my question is "yes"!

gsketefian replied:
So what do the `PCP_COMBINE` tasks do exactly for `ASNOW`?  Are the hours in the obs every 6 hours, and those get combined to 12, (not 18?), 24, (not 30?), 36, etc?  And same for forecasts?

willmayfield replied:
The obs are available as 6- and 24-hour accumulations in grib2 format from nohrsc on HPSS (and those are the accumulations recommended to be verified in the [Metrics Spreadsheet](https://docs.google.com/spreadsheets/d/1WzA2Um-kfbGs0Bh9fvFRcxzBirKyu9AQ-aHKJwAItIc/edit?usp=sharing)). Therefore those are simply retrieved and pcp_bombine_obs is not needed.  Note that the 6-h accumulations are available every 6 hours and the 24-h accumulations are available every 12 hours, so I am verifying them at that pace.

The forecasts, on the other hand, are hourly 1-h accumulations output in post. So they are run with pcp_combine_fcst to combine them into the 6- and 24-h accumulations to match the obs.

MichaelLueken commented:
@willmayfield I've some some additional testing today and it looks like incorporating some of @danielabdi-noaa's changes from PR #701 will allow both ASNOW to properly work with `FCST_LEN_HRS` of 6 and APCP to work with `FCST_LEN_HRS` of 1 and 2.  Specifically, the changes made to `ush/python_utils/environment.py` on lines 82 through 86 will allow the tests to verification tasks to properly set `ACCUM_HH`.  The verification WE2E tests were ran, along with the coverage tests for Hera Intel and GNU, and all tests passed.  Additional testing using the `MET_verification` test and setting `FCST_LEN_HRS` to 1 and 2 passed; setting `FCST_LEN_HRS` to 6 and adding ASNOW to the `MET_verification` test also passed.

willmayfield committed:
Corrected names of tasks for vx docs, and incorporated fixes for running shorter forecast lengths with 2-digit ACCUM_HH..

willmayfield committed:
Merge branch 'feature/snowfall_vx' of https://github.com/willmayfield/ufs-srweather-app into feature/snowfall_vx

willmayfield commented:
@MichaelLueken @gsketefian @gspetro-NOAA  thank you for your feedback. I fixed the typos/task names in the documentation, and incorporated the changes to ush/python_utils/environment.py that Michael suggested to solve the issue with ACCUM_HH. I also needed to add the line from ush/set_FV3nml_ens_stoch_seeds.py that is a part of @danielabdi-noaa's PR #701, in order for stochastic physics to run. With these changes the case I ran with ASNOW turned on for a 6 hour forecast was successful.

mkavulich reviewed (approved):
I tested these changes as part of my PR #864 and confirmed the snow accumulation verification gives the expected output for a winter case (2023021700).

MichaelLueken reviewed (approved):
@willmayfield Thanks for working with me to apply several of @danielabdi-noaa's bug fixes from PR #701 into this work!  Testing the current verification WE2E tests have passed, as well as testing ASNOW with a 6 hour forecast.  Approving now.

I'll go ahead and launch the automated Jenkins tests now.

Since @gsketefian noted that he is wrapping up his review of these changes, I hold off merging this PR until he has given his approval.

MichaelLueken commented:
The Cheyenne Intel tests successfully passed on Hera:

```
----------------------------------------------------------------------------------------------------
Experiment name                                                  | Status    | Core hours used 
----------------------------------------------------------------------------------------------------
custom_GFDLgrid__GFDLgrid_USE_NUM_CELLS_IN_FILENAMES_eq_FALSE      COMPLETE              13.81
grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16_plot     COMPLETE              32.38
grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16                COMPLETE              20.12
grid_RRFS_CONUScompact_13km_ics_HRRR_lbcs_RAP_suite_HRRR           COMPLETE              25.68
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta    COMPLETE               8.84
grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_HRRR_suite_HRRR                COMPLETE              16.40
pregen_grid_orog_sfc_climo                                         COMPLETE              10.12
specify_template_filenames                                         COMPLETE               9.34
----------------------------------------------------------------------------------------------------
Total                                                              COMPLETE             136.69
```
Will manually run the Cheyenne GNU tests first thing on Monday, then get this PR merged.

MichaelLueken commented:
The Cheyenne GNU tests were ran on Hera using the GNU compiler.  All tests successfully passed:
```
----------------------------------------------------------------------------------------------------
Experiment name                                                  | Status    | Core hours used 
----------------------------------------------------------------------------------------------------
grid_CONUS_25km_GFDLgrid_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16      COMPLETE              22.86
grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta      COMPLETE             236.47
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp  COMPLETE             115.17
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v17_p8_plot  COMPLETE              29.21
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_HRRR             COMPLETE              39.48
grid_RRFS_CONUScompact_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16   COMPLETE              25.17
grid_RRFS_NA_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RAP                 COMPLETE             328.18
grid_SUBCONUS_Ind_3km_ics_NAM_lbcs_NAM_suite_GFS_v16               COMPLETE              51.57
specify_EXTRN_MDL_SYSBASEDIR_ICS_LBCS                              COMPLETE              10.73
----------------------------------------------------------------------------------------------------
Total                                                              COMPLETE             858.84
```
Following either an approval or comment saying that @gsketefian is okay with the changes, this work can get merged.

gsketefian reviewed (approved):
Glad @MichaelLueken and @willmayfield figured out the `ACCUM_HH` bug.  Looks good to me and I'm approving.  I just have one suggestion.  It would be a good idea to either modify one of the WE2E tests to include snowfall accum vx or to add a new one that does this.  I've now learned to add tests consistently for features I know at least I (or projects I'm involved with) will use often.
